### PR TITLE
Make align_partitions more strict

### LINKED ISSUE #1373: DataFrame.assign behaves unpredictably with pd.Series
Calling `dataframe.assign(key=pd.Series())` causes the resulting dataframe to behave oddly. In particular, head() returns an empty dataframe and sample() is broken; however, tail() performs normally. 

For example:
```
>>> df.head()
...
[5 rows x 24 columns]

>>> df = df.assign(new_col=pd.Series(range(size)))
>>> df.head()
...
[0 rows x 25 columns]

>>> df.tail()
...
[5 rows x 25 columns]

>>> df.sample(0.1).compute()
ValueError: a must be greater than 0

Traceback
---------
  File "/.../python2.7/site-packages/dask/async.py", line 267, in execute_task
    result = _execute_task(task, data)
  File "/.../python2.7/site-packages/dask/async.py", line 249, in _execute_task
    return func(*args2)
  File "/.../python2.7/site-packages/pandas/core/generic.py", line 2572, in sample
    locs = rs.choice(axis_length, size=n, replace=replace, p=weights)
  File "mtrand.pyx", line 1391, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15395)
```

Assigning an np.Array works fine.

### PR DESCRIPTION
**Pull request by jcrist**
Previously when aligning DataFrame/Series with unknown divisions, and
ones with known divisions, bad things would occur. Now we explicitly
check if any of the input objects have unknown divisions, and fail if
they do.

A separate function used elsewhere is `_maybe_align_partitions` - this
only calls `align_partitions` if the divisions are unequal. This means
that *equal but unknown* divisions can propogate through, which is
desired for many elementwise operations. The case of mixing known and
unknown is then caught in `align_partitions`, failing as desired.

Tests for these behaviors are then added to ensure the fail cases hold.

Also:
- Cleaned up parameterization of tests in test_multi.py
- Added tests for known and unknown divisions for `assign`

Fixes #1373.

### CODE REVIEW

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -744,20 +744,40 @@ def test_index():
 
 
 def test_assign():
```
Comment (mrocklin) [inline dask/dataframe/tests/test_dataframe.py]: Should we rename this test so that if it fails we know more clearly where the problem lies?

File: dask/dataframe/tests/test_multi.py [hunk 2]
```diff
@@ -484,8 +556,9 @@ def test_join_by_index_patterns(how, shuffle):
             """
 
 
-@pytest.mark.parametrize('how', ['inner', 'outer', 'left', 'right'])
-@pytest.mark.parametrize('shuffle', ['disk', 'tasks'])
+@pytest.mark.parametrize(['how', 'shuffle'],
+                         product(['inner', 'left', 'right', 'outer'],
+                                 ['disk', 'tasks']))
```
Comment (mrocklin) [inline dask/dataframe/tests/test_multi.py]: why this change?

File: dask/dataframe/tests/test_dataframe.py (hunk 1 above)
Comment (mrocklin) [inline dask/dataframe/tests/test_dataframe.py]: Ah, I see that you have other named tests below

File: dask/dataframe/tests/test_multi.py (hunk 2 above)
Comment (jcrist) [inline dask/dataframe/tests/test_multi.py]: Changed to see match how parametrize is used elsewhere, and to see if there was any overhead to stacking them (the multi tests are by far the slowest in dataframe). Found no change, but never reverted.

Comment (mrocklin) [inline dask/dataframe/tests/test_multi.py]: I have a slight preference to keep them independent if there's no objection.

Comment (jcrist) [inline dask/dataframe/tests/test_multi.py]: Reverted.
